src/WebJobs.Script/Host/ScriptHost.cs

```diff
@@ -1089,11 +1089,13 @@ private void OnFileChanged(object sender, FileSystemEventArgs e)
             // - the host.json file was changed
             // - a function.json file was changed
             // - a function directory was added/removed
+            // - a function directory was renamed
             string fileName = Path.GetFileName(e.Name);
             if (isWatchedDirectory ||
                 ((string.Compare(fileName, ScriptConstants.HostMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
                 string.Compare(fileName, ScriptConstants.FunctionMetadataFileName, StringComparison.OrdinalIgnoreCase) == 0) ||
-                (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot))
+                (Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).Count() != _directoryCountSnapshot) ||
+                (e.ChangeType == WatcherChangeTypes.Renamed && String.IsNullOrEmpty(Path.GetExtension(e.FullPath))))
```

Comment:
The actual behavior would depend on the implementation/what gets returned. In this case, it may indeed cause problems. 

To avoid problems and confusion, you can either call `ToList`/`ToArray` to materialize the results. You could still assign to `IEnumerable<string>` (which would be materialized, just a List or array at that point) but to make it clearer, it would probably be a good idea to make the type match what you're returning, or ideally, just have the actual snapshot be an `ImmutableCollection<string>` created from the result of the `EnumerateDirectories` call.